test/MiddlewareControllerTest.php

```diff
@@ -0,0 +1,158 @@
+ */
+
+namespace ZendTest\Mvc\Controller;
+
+use PHPUnit\Framework\TestCase;
+use Psr\Http\Message\ResponseInterface;
+use Zend\EventManager\EventManager;
+use Zend\EventManager\EventManagerInterface;
+use Zend\Http\Request;
+use Zend\Http\Response;
+use Zend\Mvc\Controller\AbstractController;
+use Zend\Mvc\Controller\MiddlewareController as DeprecatedMiddlewareController;
+use Zend\Mvc\Middleware\MiddlewareController;
+use Zend\Mvc\Exception\RuntimeException;
+use Zend\Mvc\MvcEvent;
+use Zend\Stdlib\DispatchableInterface;
+use Zend\Stdlib\RequestInterface;
+use Zend\Stratigility\MiddlewarePipe;
+
+/**
+ * @covers \Zend\Mvc\Middleware\MiddlewareController
+ */
+class MiddlewareControllerTest extends TestCase
+{
+    /**
+     * @var MiddlewarePipe|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $pipe;
+
+    /**
+     * @var ResponseInterface|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $responsePrototype;
+
+    /**
+     * @var EventManagerInterface
+     */
+    private $eventManager;
+
+    /**
+     * @var AbstractController|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $controller;
+
+    /**
+     * @var MvcEvent
+     */
+    private $event;
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function setUp()
+    {
+        $this->pipe              = $this->createMock(MiddlewarePipe::class);
+        $this->responsePrototype = $this->createMock(ResponseInterface::class);
+        $this->eventManager      = $this->createMock(EventManagerInterface::class);
+        $this->event             = new MvcEvent();
+        $this->eventManager      = new EventManager();
+
+        $this->controller = new MiddlewareController(
+            $this->pipe,
+            $this->responsePrototype,
+            $this->eventManager,
+            $this->event
+        );
+    }
+
+    public function testEMHasDeprecatedControllerIdentifier()
+    {
+        $controller = new MiddlewareController(
+            $this->pipe,
+            $this->responsePrototype,
+            new EventManager(),
+            $this->event
+        );
+        $identifiers = $controller->getEventManager()->getIdentifiers();
+        $this->assertContains(DeprecatedMiddlewareController::class, $identifiers);
+    }
+
+    public function testWillAssignCorrectEventManagerIdentifiers()
+    {
+        $identifiers = $this->eventManager->getIdentifiers();
+
+        self::assertContains(MiddlewareController::class, $identifiers);
+        self::assertContains(AbstractController::class, $identifiers);
+        self::assertContains(DispatchableInterface::class, $identifiers);
+    }
+
+    public function testWillDispatchARequestAndResponseWithAGivenPipe()
+    {
+        $request          = new Request();
+        $response         = new Response();
+        $result           = $this->createMock(ResponseInterface::class);
+        /* @var $dispatchListener callable|\PHPUnit_Framework_MockObject_MockObject */
```

Comment:
`@var paramType $paramName`

not sure also if `PHPUnit_Framework_MockObject_MockObject` is right for all used PHPUnit versions